### PR TITLE
feat(claims): add runtime-scoped claim writes

### PR DESCRIPTION
**Pull request by jonathanhaaswriter**
## Summary
- add generic Claim and EntityRef protos plus BootstrapService WriteClaims
- persist runtime-scoped claims in Postgres and mirror entity refs and asserted relation claims into projections
- add the builtin sdk source and TS/Python SDK helpers for push-based app onboarding

## Testing
- make verify
- python3 -m py_compile sdk/python/cerebro_sdk/client.py sdk/python/cerebro_sdk/__init__.py
- cd sdk/typescript && npx -y -p typescript tsc -p tsconfig.json

### CODE REVIEW

**Comment by jonathanhaaswriter** [comment]
@droid Please perform a deep code review and security review of this PR. Leave inline comments only for high-confidence, actionable issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review this PR after the latest fixes. Focus on whether the prior code review and security findings are resolved, and call out any remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
@droid Please re-review after the latest pushed fixes. Focus on whether the prior findings on this PR are resolved and only report remaining high-confidence issues.

**Comment by jonathanhaaswriter** [comment]
### Droid sweep status

Latest sweep: `2026-04-29 22:09 UTC`.

- Unresolved non-outdated review threads: `0`
- CI `verify`: `SUCCESS`
- `droid-review`: `not present`

I’ll keep watching for new review findings/check failures and address them as they appear.

**Comment by jonathanhaaswriter** [comment]
Re-reviewed the still-open outdated thread(s) on the latest branch tip. The branch already includes object identity in generated claim IDs, scopes the Postgres claims key by tenant/runtime, and URL-encodes `runtime_id` in the Python client, so no code change was needed. Validation: `go test ./internal/claims`. Remaining assigned unresolved threads after this audit: none.

**Comment by jonathanhaaswriter** [comment]
Droid sweep audit complete (final pass)

- Re-scanned all live review threads, including outdated ones
- Verified each remaining thread against the current branch tip
- Resolved every actionable thread (or fixed it inline and pushed)
- 0 unresolved review threads across the whole open-PR stack at this time

PR #494 (fix/feedback-main-followup) and PR #495 (feat/droid-creation-workflows) received additional fixes for SSRF (DNS-rebind) and workflow branch-name validation in this pass.

**Comment by jonathanhaaswriter** [comment]
@droid review

**Comment by jonathanhaaswriter** [comment]
@droid please harden this PR end-to-end:

1. Re-review the diff for residual bugs, races, edge cases, security gaps, and untested paths.
2. Apply the smallest correct fixes for any high-confidence issues you find.
3. Strengthen test coverage: add regression tests for every prior review finding on this PR plus the new behavior, including failure paths, boundary inputs, concurrency, and not-found / typed-nil interface guards.
4. Run focused tests and `make verify`. If validators fail, fix until green.
5. Commit and push the fixes to this PR's branch (or, if you lack push permission in this workflow, leave concrete inline suggestions and a summary comment).

Do not change the PR scope, do not modify the default branch, and do not merge.

**Comment by jonathanhaaswriter** [comment]
@droid review

File: internal/claims/service.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+		return nil, errors.New("source runtime is required")
+	}
+	normalized := proto.Clone(claim).(*cerebrov1.Claim)
+	normalized.SubjectUrn = strings.TrimSpace(normalized.GetSubjectUrn())
+	normalized.Predicate = strings.TrimSpace(normalized.GetPredicate())
+	normalized.ObjectUrn = strings.TrimSpace(normalized.GetObjectUrn())
+	normalized.ObjectValue = strings.TrimSpace(normalized.GetObjectValue())
+	normalized.ClaimType = strings.TrimSpace(normalized.GetClaimType())
+	normalized.Status = strings.TrimSpace(normalized.GetStatus())
+	normalized.SourceEventId = strings.TrimSpace(normalized.GetSourceEventId())
```
[the remark this answers is not in the document]
Comment (jonathanhaaswriter) [reply internal/claims/service.go:180]: Fixed in 4606ec624 by trimming `Claim.Id` before the empty-ID check so whitespace IDs get deterministic generated IDs; added a regression.

**Comment by jonathanhaaswriter** [comment]
@droid review

Please re-review the runtime-scoped claim writes deeply. Generic claim writes are a powerful API surface that needs strong validation:

1. **Schema and proto safety**: verify `Claim` and `EntityRef` shapes reject ambiguous/empty entity refs, malformed URNs, oversized fields, and that proto evolution remains backward-compatible for SDKs.
2. **Tenant/runtime isolation**: ensure WriteClaims always scopes inserts and projection mirrors to the caller’s runtime/tenant, and that neither the bootstrap surface nor the builtin sdk source can be coerced into writing across tenants.
3. **Authorization on entity refs**: confirm callers cannot write claims referencing entities they don’t own (e.g., another tenant’s URNs or platform-owned nodes), and that validation runs before any persistence.
4. **Idempotency and conflict semantics**: re-check upsert keys, repeated writes, concurrent writers, and how partial Postgres failures surface; verify projection mirrors stay consistent with persisted claim batches.
5. **Input bounds and DoS**: ensure batch size, per-claim size, attribute count, and string length limits are enforced as `InvalidArgument`, with no path to amplify a single request into an unbounded write.
6. **SDK helper correctness**: verify Python and TypeScript helpers normalize inputs identically (e.g., null/empty handling, default booleans, scope/list tolerance) and produce equivalent on-the-wire claims.
7. **Forward-compat with reads/reconciliation**: confirm the write contract supports filtered reads (PR 464) and snapshot reconciliation (PR 474) without future breaking changes to keys, fingerprints, or projection mirrors.

**Comment by jonathanhaaswriter** [comment]
Closing this PR — its substantive feature work is already in `main` via the earlier consolidation rollups, and the unresolved factory-droid findings posted on this PR have been addressed in main via the follow-up PRs:

- #497 (`fix(github): drop non-cursor opaque from terminal audit checkpoints`)
- #499 (`fix: address post-merge feedback across pending stacks`)

PR #499 specifically addresses the findings posted on this stack:

- Reject malformed `root_urn` values in `GetEntityNeighborhood` (PR #460)
- Map `refuted` and `superseded` claim statuses to projection deletion (PR #463)
- Tenant-scope `ListClaims` reads (PR #464)
- Per-runtime serialization of `WriteClaims` (PR #474)
- Restore null-array normalization in `graph_summary` (PR #470)
- Surface blank-root errors from `graphLayering` (PR #468)

This branch has diverged ~50k lines from `main` and rebasing it cleanly is no longer practical; closing here keeps the open queue accurate. If anything from this stack still needs to land, please open a fresh PR cut from `main`.
